[WIP] Server/Tools: IRCBridge

ccrs opened the pull request:
Re-implementation of TrinityChat (#19376).

Threaded tool that connects to the target IRC server, sends triggered information and listens for logged in users commands.

(removed in 2265aef916e)

ccrs committed:
placeholder

ccrs committed:
configuration support

ccrs committed:
missing logger assignation on config file

Shauren commented on src/server/worldserver/IRCBridge/IRCBridge.h:
```diff
@@ -0,0 +1,92 @@
+    CONFIGURATION_IRCBRIDGE_STRING_COUNT
+};
+
+class IRCBridgeConnectionHandler
```
You really should just use our `Socket` class to handle this

ccrs replied:
oh, will take a look into it

Shauren replied:
you might need to still implement the connect part tho

jackpoz commented:
> Re-implementation of TrinityChat

do you mean you fixed all the issues of the previous PR or is this whole new code (so we need to review it all over again) ?

tkrokli commented on src/server/worldserver/IRCBridge/IRCBridge.cpp:
```diff
@@ -0,0 +1,75 @@
+    return true;
+}
+
+uint32 IRCBridge::LoadConfiguration(IRCBridgeConfigurationUintValues index, char const * fieldname, uint32 defvalue) const
```
```
src/server/worldserver/IRCBridge/IRCBridge.cpp:53:70: fatal error: 
      unused parameter 'index' [-Wunused-parameter]
uint32 IRCBridge::LoadConfiguration(IRCBridgeConfigurationUintValues index, ...
                                                                     ^
```

ccrs replied:
thanks!

ccrs commented:
@jackpoz pretty much yes, the other PR is just something we better forget about and only use its concept haha

ccrs committed:
initial implementation of asynchronous connection & loadconfiguration correction

ccrs committed:
followup

ccrs committed:
fix build

tkrokli commented on src/server/worldserver/IRCBridge/IRCBridge.cpp:
```diff
@@ -0,0 +1,169 @@
+#include "World.h"
+#include "IRCBridge.h"
+
+#include <boost\asio\connect.hpp>
```
```
src/server/worldserver/IRCBridge/IRCBridge.cpp:6:10: fatal error: 
      'boost\asio\connect.hpp' file not found
#include <boost\asio\connect.hpp>
         ^
```

ccrs committed:
testing

ccrs committed:
use a static call, fix crash and fix build

ccrs committed:
back to the member function

ccrs committed:
fix warning

ccrs committed:
first implementation of the handler

ccrs committed:
fix warning

ccrs committed:
minor corrections

ccrs committed:
more corrections

ccrs committed:
another correction

ccrs committed:
even more corrections

tkrokli commented:
Maybe the CMakeFiles.txt files need to be checked, or is this error caused by something else?
```
CMakeFiles/worldserver.dir/IRCBridge/IRCBridge.cpp.o: In function `IRCBridge::Run()':
/src/server/worldserver/IRCBridge/IRCBridge.cpp:(.text+0x6db): undefined reference to `IRCBridgeHandler::instance()'
/src/server/worldserver/IRCBridge/IRCBridge.cpp:(.text+0x6e6): undefined reference to `IRCBridgeHandler::GetNext()'
clang: fatal error: linker command failed with exit code 1 (use -v to see invocation)
make[2]: *** [src/server/worldserver/worldserver] Error 1
make[2]: Target `src/server/worldserver/CMakeFiles/worldserver.dir/build' not remade because of errors.
make[1]: *** [src/server/worldserver/CMakeFiles/worldserver.dir/all] Error 2
make[1]: Target `all' not remade because of errors.
make: *** [all] Error 2
make: Target `default_target' not remade because of errors.
```

ccrs committed:
move socket to its own file & fix reading

ccrs committed:
handle ping - pong

ccrs commented:
We have connection boyz

@tkrokli I dont know whats causing that

tkrokli commented:
Sorry, my bad. I did not read the clang error message properly. It looks like clang expects to find references to `IRCBridgeHandler::instance()` and `IRCBridgeHandler::GetNext()` when reading `IRCBridge::Run()` in the file `IRCBridge.cpp` . Finding the reference could be achieved if those 2 were defined in `IRCBridgeHandler.h`, but they are only defined in `IRCBridgeHandler.cpp`. Maybe it could be worth looking into, I'm not sure.

ccrs committed:
dynamic linking

ccrs committed:
fix shutdown & fix crash on shutdown & remove unneded header & create socket with master reference

ccrs committed:
correct stop

ccrs committed:
missing header

ccrs committed:
handle error & implement status & handle connection code as string & implement connection code as login callback

ccrs committed:
use char on string find & log the rest of the message to debug

ccrs committed:
litle reorder

ccrs committed:
little reorder

ccrs committed:
use boost for string transform

Rochet2 commented on src/server/worldserver/IRCBridge/IRCBridge.cpp:
```diff
@@ -256,6 +257,7 @@ void IRCBridge::Login()
-    return string;
+    std::string temp = string;
+    boost::to_upper(temp);
+    return temp;
```
Why change to boost? Less dependecy without it.
Also the "string" is already a copy when passed, so no need to make temp

ccrs replied:
there is no return value, so I have to copy again

and boost over std::transform because its more efficient and has less troubles with special chars

ccrs replied:
confused? elaborate

ccrs committed:
feedback

ccrs committed:
remove unneded field on command & implement reconnect timer

ccrs committed:
some reorder

ccrs committed:
handle reconnect on connection error

ccrs committed:
implement connection attempts

ccrs committed:
use timetracker

ccrs committed:
rename handler (going to be generic)

ccrs committed:
relay target types & move reports to handler (bridge wont use it internally) & prevent loop ticks while reconnecting

ccrs committed:
couple logs & bind logs also to console

tkrokli commented on src/server/worldserver/IRCBridge/IRCBridge.h:
```diff
@@ -0,0 +1,124 @@
+        std::thread _thread;
+
+        IRCBridgeStatus _status;
+        RelayTargetType _relayType;
```
```
In file included from /src/server/worldserver/IRCBridge/IRCBridge.cpp:7:
/src/server/worldserver/IRCBridge/IRCBridge.h:116:25: fatal error: 
      private field '_relayType' is not used [-Wunused-private-field]
        RelayTargetType _relayType;
                        ^
```

Shauren replied:
@tkrokli please don't comment about warnings in pull requests clearly labeled as WIP

ccrs replied:
this time its true, its unnused because I still have not finished working on it

ccrs committed:
handle RelayTargetType

ccrs committed:
semaphore the income/outcome of data from the handler

ccrs committed:
change to deque

std::queue is a type of container adaptor, used as underlying container; vector or deque is selected if no class is specified

ccrs committed:
create IsEnabled helper

ccrs committed:
fix shutdown crash

raelik commented:
Trying to test this on Linux, doesn't send anything to the IRC server after connection.  Appears to receive just fine however.

raelik commented on src/server/worldserver/IRCBridge/IRCBridge.cpp:
```diff
@@ -0,0 +1,294 @@
+
+    _active = false;
+
+    if (_thread.joinable())
```
I'm pretty sure this is intentional, as this is in the destructor and is part of the shutdown code. It only joins on the thread if it is alive and not already joined.

Aokromes commented:
Conflicting files
src/server/worldserver/Main.cpp

Aokromes commented on src/server/game/Tools/RelayHandler/RelayCommand.cpp:
```diff
@@ -0,0 +1 @@
+#include "RelayCommand.h"
```
missing header :P

Aokromes commented on src/server/game/Tools/RelayHandler/RelayHandler.cpp:
```diff
@@ -0,0 +1,54 @@
+#include "RelayHandler.h"
```
missing header :P

Aokromes commented on src/server/game/Tools/RelayHandler/RelayScripts.cpp:
```diff
@@ -0,0 +1 @@
+#include "RelayScripts.h"
```
missing header :P

Aokromes commented on src/server/worldserver/IRCBridge/IRCBridge.cpp:
```diff
@@ -0,0 +1,294 @@
+#include "Common.h"
```
missing header :P

Aokromes commented on src/server/worldserver/IRCBridge/IRCBridgeSocket.cpp:
```diff
@@ -0,0 +1,92 @@
+#include "Log.h"
```
missing header :P

Aokromes commented on src/server/worldserver/IRCBridge/IRCBridgeSocket.h:
```diff
@@ -0,0 +1,54 @@
+/*
+ * Copyright (C) 2008-2017 TrinityCore <http://www.trinitycore.org/>
```
update (C)

tkrokli commented on src/server/scripts/Commands/cs_relay.cpp:
```diff
@@ -0,0 +1,55 @@
+        return commandTable;
+    }
+
+    static bool HandleRelaySendCommand(ChatHandler* /*handler*/, char const* args)
```
```cpp
src/server/scripts/Commands/cs_relay.cpp:45:40:
 fatal error: unknown type name 'ChatHandler'
    static bool HandleRelaySendCommand(ChatHandler* /*handler*/, char const* args)
                                       ^
1 error generated.
```

ccrs commented on src/server/worldserver/IRCBridge/IRCBridge.cpp:
```diff
@@ -0,0 +1,294 @@
+
+    _active = false;
+
+    if (_thread.joinable())
```
^ that

jackpoz commented:
Feel free to ask to reopen this PR if you plan to resume the development
